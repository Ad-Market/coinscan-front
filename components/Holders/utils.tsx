import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import isoWeek from 'dayjs/plugin/isoWeek';

import {
  DailyHodlersStatesDocument,
  GetTransactionsPaginatedQuery,
} from '../../generated/graphql';

dayjs.extend(isoWeek);
dayjs.extend(utc);

type timeFrameFn = (date: Date) => Date;

export type TransactionsQueryData =
  GetTransactionsPaginatedQuery['transactions'];

export type DataPoint = {
  id: string;
  count: number;
};

export const currentTZ = Intl.DateTimeFormat().resolvedOptions().timeZone;

export const getUnixTime = (date: Date): number =>
  Math.floor(date.getTime() / 1000);

export const fromUnixTime = (unixTime: number): Date =>
  new Date(unixTime * 1000);

export const toLocaleDateStringUTC = (date: Date): string =>
  dayjs(date).utc().toDate().toLocaleDateString(undefined, { timeZone: 'UTC' });

export const toLocaleStringUTC = (date: Date): string =>
  dayjs(date).utc().toDate().toLocaleString(undefined, { timeZone: 'UTC' });

export const formatUTC = (date: Date, format: string): string =>
  dayjs(date).utc().format(format);

export const startOfDay = (date: Date): Date =>
  dayjs(date).utc().startOf('day').toDate();

export const startOfNextDay = (date: Date): Date =>
  dayjs(date).utc().add(1, 'day').startOf('day').toDate();

export const endOfWeek = (date: Date): Date =>
  dayjs(date).utc().endOf('isoWeek').toDate();

export const previousMonday = (date: Date): Date =>
  dayjs(date).utc().startOf('isoWeek').toDate();

export const nextMonday = (date: Date): Date => {
  return dayjs(date).utc().add(7, 'day').startOf('isoWeek').toDate();
};

export const startOfMonth = (date: Date): Date => {
  return dayjs(date).utc().startOf('month').toDate();
};

export const startOfNextMonth = (date: Date): Date => {
  return dayjs(date).utc().add(1, 'month').startOf('month').toDate();
};

const groupDataByTimeframe = (
  data: DataPoint[],
  currentTimeframeFn: timeFrameFn,
  nextTimeframeFn: timeFrameFn,
  aggregator: typeof maxAggregator
): DataPoint[] => {
  if (!Array.isArray(data) || data.length === 0) {
    return data;
  }
  let groupedData = [...data].reverse();

  const firstDate = parseInt(groupedData[0].id, 10);

  let currentTimeframeDate = currentTimeframeFn(fromUnixTime(firstDate));
  let nextTimeframeDate = nextTimeframeFn(fromUnixTime(firstDate));

  groupedData = groupedData.reduce((acc: DataPoint[], dataItem) => {
    if (acc.length === 0) {
      acc.push({
        id: getUnixTime(currentTimeframeDate).toString(),
        count: dataItem.count,
      });
      return acc;
    }

    if (parseInt(dataItem.id, 10) > getUnixTime(nextTimeframeDate)) {
      currentTimeframeDate = currentTimeframeFn(
        fromUnixTime(parseInt(dataItem.id, 10))
      );
      nextTimeframeDate = nextTimeframeFn(currentTimeframeDate);
      acc.push({
        id: getUnixTime(currentTimeframeDate).toString(),
        count: dataItem.count,
      });
    } else {
      const lastAccItemIndex = acc.length - 1;
      acc[lastAccItemIndex].count = aggregator(acc, lastAccItemIndex, dataItem);
    }
    return acc;
  }, []);
  return groupedData.reverse();
};

const maxAggregator = (
  acc: DataPoint[],
  lastAccItemIndex: number,
  dataItem: DataPoint
) => Math.max(dataItem.count, acc[lastAccItemIndex].count);

const sumAggregator = (
  acc: DataPoint[],
  lastAccItemIndex: number,
  dataItem: DataPoint
) => dataItem.count + acc[lastAccItemIndex].count;

export const groupDataSumByDays = (data: DataPoint[]) =>
  groupDataByTimeframe(data, startOfDay, startOfNextDay, sumAggregator);

export const groupDataMaxByWeeks = (data: DataPoint[]) =>
  groupDataByTimeframe(data, previousMonday, nextMonday, maxAggregator);

export const groupDataMaxByMonths = (data: DataPoint[]) =>
  groupDataByTimeframe(data, startOfMonth, startOfNextMonth, maxAggregator);

export const convertTransactionsArrayToDataPointArray = (
  transactions: TransactionsQueryData,
  relativeTowallet: string
): DataPoint[] => {
  return transactions.map((transaction) => ({
    id: transaction.timestamp,
    count:
      transaction.from.address === relativeTowallet
        ? -transaction.value
        : +transaction.value,
  }));
};

export const fillMissingDaysInDataPointArray = (
  data: DataPoint[],
  days: number
) => {
  const filledData: DataPoint[] = [];
  const today = dayjs.utc();
  for (
    let d = today.startOf('day').subtract(days - 1, 'day');
    d <= today.endOf('day');
    d = d.add(1, 'day')
  ) {
    filledData.push({
      id: getUnixTime(d.startOf('day').toDate()).toString(),
      count: 0,
    });
  }

  data.forEach((dataPoint) => {
    const filledDataPoint = filledData.find((d) => d.id === dataPoint.id);
    if (filledDataPoint) {
      filledDataPoint.count = dataPoint.count;
    }
  });
  return filledData;
};
