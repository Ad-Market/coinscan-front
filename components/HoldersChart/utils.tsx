import {
  fromUnixTime,
  previousMonday,
  nextMonday,
  getUnixTime,
  startOfMonth,
  addMonths,
} from 'date-fns';
import { DailyHodlersStatesQuery } from '../../generated/graphql';

const startOfNextMonth = (date: Date) => startOfMonth(addMonths(date, 1));

type DataPoint = {
  id: string;
  count: number;
};

type timeFrameFn =
  | typeof previousMonday
  | typeof nextMonday
  | typeof startOfMonth
  | typeof startOfNextMonth;

export type ChartData = DailyHodlersStatesQuery['dailyHoldersStates'];

const groupDataByMaxInTimeframe = (
  data: ChartData,
  currentTimeframeFn: timeFrameFn,
  nextTimeframeFn: timeFrameFn
): ChartData => {
  if (!Array.isArray(data) || data.length === 0) {
    return data;
  }
  let groupedData = [...data].reverse();

  const firstDate = fromUnixTime(parseInt(groupedData[0].id, 10));
  let currentTimeframeDate = currentTimeframeFn(firstDate);
  let nextTimeframeDate = nextTimeframeFn(firstDate);

  groupedData = groupedData.reduce((acc: DataPoint[], dataItem) => {
    if (acc.length === 0) {
      acc.push({
        id: getUnixTime(currentTimeframeDate).toString(),
        count: dataItem.count,
      });
    }

    if (fromUnixTime(parseInt(dataItem.id, 10)) > nextTimeframeDate) {
      currentTimeframeDate = nextTimeframeDate;
      nextTimeframeDate = nextTimeframeFn(currentTimeframeDate);
      acc.push({
        id: getUnixTime(currentTimeframeDate).toString(),
        count: dataItem.count,
      });
    } else {
      const lastAccItemIndex = acc.length - 1;
      // update value to current if larger
      acc[lastAccItemIndex].count = Math.max(
        dataItem.count,
        acc[lastAccItemIndex].count
      );
    }
    return acc;
  }, []);
  return groupedData.reverse();
};

export const groupDataMaxByWeeks = (data: ChartData) =>
  groupDataByMaxInTimeframe(data, previousMonday, nextMonday);

export const groupDataMaxByMonths = (data: ChartData) =>
  groupDataByMaxInTimeframe(data, startOfMonth, startOfNextMonth);
