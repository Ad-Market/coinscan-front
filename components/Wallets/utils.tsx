import type { GetWalletsPaginatedWithTransactionsQuery } from '../../generated/graphql';

export type Wallet = GetWalletsPaginatedWithTransactionsQuery['wallets'][0];

export function getNetFlowPercentageFromWallet(
  wallet: Wallet | null | undefined,
  timestamp: number
): number {
  let percent = 0;
  if (!wallet) return percent;

  const positiveFlow = wallet.dailyStates.reduce((acc, dailyState) => {
    if (Number(dailyState.start) >= timestamp) {
      return acc + BigInt(dailyState.inflow);
    }
    return acc;
  }, BigInt(0));

  const negativeFlow = wallet.dailyStates.reduce((acc, dailyState) => {
    if (Number(dailyState.start) >= timestamp) {
      return acc + BigInt(dailyState.outflow);
    }
    return acc;
  }, BigInt(0));

  const previousWalletBalance =
    BigInt(wallet.value) - positiveFlow + negativeFlow;

  // clamp bar width to max 100% (show real value > 100% only as text)
  if (previousWalletBalance === BigInt(0)) {
    if (BigInt(wallet.value) > BigInt(0)) {
      percent = 100;
    } else {
      percent = 0;
    }
  } else {
    percent =
      Number(
        ((positiveFlow - negativeFlow) * BigInt(10000)) / previousWalletBalance
      ) / 100;
  }

  return percent;
}
