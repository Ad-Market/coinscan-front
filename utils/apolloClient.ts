import { ApolloClient, ApolloLink, InMemoryCache } from '@apollo/client';
import { HttpLink } from '@apollo/client/link/http';
import {
  ErrorHandler,
  onError as apolloClientOnError,
} from '@apollo/client/link/error';
import settings from '../settings.json';

export const paginationMerge = (existing: any, incoming: any) => {
  return [...incoming];
};

export const createApolloClient = (onError: ErrorHandler = () => {}) => {
  const httpLink = new HttpLink({
    uri: settings.graphqlUri,
  });
  const errorLink = apolloClientOnError(onError);

  return new ApolloClient({
    link: ApolloLink.from([errorLink, httpLink]),
    defaultOptions: {
      query: { errorPolicy: 'ignore' },
    },
    cache: new InMemoryCache({
      typePolicies: {
        Query: {
          fields: {
            walletTransactions: {
              keyArgs: false,
              merge: paginationMerge,
            },
            wallets: {
              keyArgs: false,
              merge: paginationMerge,
            },
            transactions: {
              keyArgs: false,
              merge: paginationMerge,
            },
          },
        },
      },
    }),
  });
};

export default createApolloClient;
