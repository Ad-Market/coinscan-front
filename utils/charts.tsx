import dayjs from 'dayjs';
import { XAxisProps } from 'recharts';

import {
  DailyWalletState,
  GetTransactionsPaginatedQuery,
} from '../generated/graphql';
import settings from '../settings.json';
import { Props as DefaultTooltipContentProps } from 'recharts/types/component/DefaultTooltipContent';
import {
  fromUnixTime,
  getUnixTime,
  startOfDay,
  startOfNextDay,
  previousMonday,
  nextMonday,
  startOfMonth,
  startOfNextMonth,
  toLocaleDateStringUTC,
  formatUTC,
  endOfWeek,
} from './time';

type timeFrameFn = (date: Date) => Date;

export type TransactionsQueryData =
  GetTransactionsPaginatedQuery['transactions'];

export interface DataPoint<T> {
  id: string;
  count: T;
}

export interface DataPointWithDisplay<T> extends Omit<DataPoint<T>, 'count'> {
  count: number;
  display: bigint;
}

type Numberish = number | bigint;

const groupDataByTimeframe = <T extends Numberish>(
  data: DataPoint<T>[],
  currentTimeframeFn: timeFrameFn,
  nextTimeframeFn: timeFrameFn,
  aggregator: typeof maxAggregator
): DataPoint<T>[] => {
  if (!Array.isArray(data) || data.length === 0) {
    return data;
  }
  let groupedData = [...data].reverse();

  const firstDate = parseInt(groupedData[0].id, 10);

  let currentTimeframeDate = currentTimeframeFn(fromUnixTime(firstDate));
  let nextTimeframeDate = nextTimeframeFn(fromUnixTime(firstDate));

  groupedData = groupedData.reduce(
    (acc: DataPoint<T>[], dataItem: DataPoint<T>) => {
      if (acc.length === 0) {
        acc.push({
          id: getUnixTime(currentTimeframeDate).toString(),
          count: dataItem.count,
        });
        return acc;
      }

      if (parseInt(dataItem.id, 10) > getUnixTime(nextTimeframeDate)) {
        currentTimeframeDate = currentTimeframeFn(
          fromUnixTime(parseInt(dataItem.id, 10))
        );
        nextTimeframeDate = nextTimeframeFn(currentTimeframeDate);
        acc.push({
          id: getUnixTime(currentTimeframeDate).toString(),
          count: dataItem.count,
        });
      } else {
        const lastAccItemIndex = acc.length - 1;
        acc[lastAccItemIndex].count = aggregator(
          acc,
          lastAccItemIndex,
          dataItem
        );
      }
      return acc;
    },
    []
  );
  return groupedData.reverse();
};

const maxAggregator = <T extends Numberish>(
  acc: DataPoint<T>[],
  lastAccItemIndex: number,
  dataItem: DataPoint<T>
): T => {
  if (dataItem.count > acc[lastAccItemIndex].count) {
    return dataItem.count;
  }
  if (dataItem.count < acc[lastAccItemIndex].count) {
    return acc[lastAccItemIndex].count;
  }

  return dataItem.count;
};

type ISumAggregator = {
  (
    acc: DataPoint<bigint>[],
    lastAccItemIndex: number,
    dataItem: DataPoint<bigint>
  ): bigint;
  (
    acc: DataPoint<number>[],
    lastAccItemIndex: number,
    dataItem: DataPoint<number>
  ): number;
};

const sumAggregator: ISumAggregator = (
  acc: any,
  lastAccItemIndex: number,
  dataItem: any
): any => dataItem.count + acc[lastAccItemIndex].count;

export const groupDataSumByDays = <T extends Numberish>(data: DataPoint<T>[]) =>
  groupDataByTimeframe(data, startOfDay, startOfNextDay, sumAggregator);

export const groupDataMaxByWeeks = <T extends Numberish>(
  data: DataPoint<T>[]
) => groupDataByTimeframe(data, previousMonday, nextMonday, maxAggregator);

export const groupDataMaxByMonths = <T extends Numberish>(
  data: DataPoint<T>[]
) => groupDataByTimeframe(data, startOfMonth, startOfNextMonth, maxAggregator);

export const convertTransactionsArrayToDataPointArray = (
  transactions: TransactionsQueryData,
  relativeTowallet: string
): DataPoint<bigint>[] => {
  const dataPointArray: DataPoint<bigint>[] = transactions.map(
    (transaction): DataPoint<bigint> => ({
      id: transaction.timestamp,
      count:
        transaction.from.address === relativeTowallet
          ? BigInt(0) - BigInt(transaction.value)
          : BigInt(transaction.value),
    })
  );
  return dataPointArray;
};

export const convertWalletDailyStatesToDataPointArray = (
  dailyStates: Pick<DailyWalletState, 'start' | 'inflow' | 'outflow'>[]
): DataPoint<bigint>[] => {
  return dailyStates.map((dailyState) => ({
    id: dailyState.start,
    count: BigInt(dailyState.inflow) - BigInt(dailyState.outflow),
  }));
};

export const fillMissingDaysInDataPointArray = (
  data: DataPoint<bigint>[],
  days: number
): DataPoint<bigint>[] => {
  const filledData: DataPoint<bigint>[] = [];
  const today = dayjs.utc();
  for (
    let d = today.startOf('day').subtract(days - 1, 'day');
    d <= today.endOf('day');
    d = d.add(1, 'day')
  ) {
    filledData.push({
      id: getUnixTime(d.startOf('day').toDate()).toString(),
      count: BigInt(0),
    });
  }

  data.forEach((dataPoint) => {
    const filledDataPoint = filledData.find((d) => d.id === dataPoint.id);
    if (filledDataPoint) {
      filledDataPoint.count = dataPoint.count;
    }
  });
  return filledData;
};

export const calculateHistoryBalanceFromTransactions = (
  transactions: DataPoint<bigint>[],
  balance: bigint
): DataPoint<bigint>[] => {
  let currentBalance = BigInt(balance);
  transactions.reverse();
  let balanceHistory: DataPoint<bigint>[] = [];

  transactions.forEach((transaction) => {
    currentBalance = BigInt(currentBalance) - BigInt(transaction.count);
    balanceHistory = [
      {
        id: transaction.id,
        count: currentBalance,
      },
      ...balanceHistory,
    ];
  });
  return balanceHistory;
};

export const convertBigIntToNumberWithoutDecimalPlacesPrecision = (
  value: bigint
): number => {
  const dataMaxString = value.toString();
  return Number(
    dataMaxString.substring(0, dataMaxString.length - settings.decimalPlaces)
  );
};

export const convertToChartableData = (
  data: DataPoint<bigint>[]
): DataPointWithDisplay<bigint>[] => {
  const displayableData = data.map((dataP) => ({
    ...dataP,
    count: convertBigIntToNumberWithoutDecimalPlacesPrecision(dataP.count),
    display: dataP.count,
  }));

  return displayableData;
};

export const formatMin = (
  dataMin: Numberish,
  roundingBase: number = Math.pow(10, settings.decimalPlaces)
): number => {
  if (typeof dataMin === 'bigint') {
    dataMin = convertBigIntToNumberWithoutDecimalPlacesPrecision(dataMin);
  }
  return dataMin - (dataMin % roundingBase);
};

export const formatMax = (
  dataMax: Numberish,
  roundingBase: number = Math.pow(10, settings.decimalPlaces)
): number => {
  if (typeof dataMax === 'bigint') {
    dataMax = convertBigIntToNumberWithoutDecimalPlacesPrecision(dataMax);
  }
  return dataMax + (roundingBase - (dataMax % roundingBase));
};

export enum ChartTimeGroupings {
  BY_DAY,
  BY_WEEK,
  BY_MONTH,
}

export const LINE_CHART_GROUPS = [ChartTimeGroupings.BY_DAY];
export const BAR_CHART_GROUPS = [
  ChartTimeGroupings.BY_WEEK,
  ChartTimeGroupings.BY_MONTH,
];

type XAxisTickFormatterFn = XAxisProps['tickFormatter'];

type XAxisTickFormatters = {
  [key in ChartTimeGroupings]: XAxisTickFormatterFn;
};

export const CHART_TIME_XAXIS_TICK_FORMATTERS: XAxisTickFormatters = {
  [ChartTimeGroupings.BY_DAY]: (value) =>
    toLocaleDateStringUTC(fromUnixTime(parseInt(value, 10))),
  [ChartTimeGroupings.BY_WEEK]: (value) =>
    toLocaleDateStringUTC(fromUnixTime(parseInt(value, 10))),
  [ChartTimeGroupings.BY_MONTH]: (value) =>
    formatUTC(fromUnixTime(parseInt(value, 10)), 'MM.YYYY'),
};

type TooltipLabelFormatterFn = DefaultTooltipContentProps<
  string,
  number
>['labelFormatter'];

type TooltipLabelFormatters = {
  [key in ChartTimeGroupings]: TooltipLabelFormatterFn;
};

export const CHART_TIME_TOOLTIP_LABEL_FORMATTERS: TooltipLabelFormatters = {
  [ChartTimeGroupings.BY_DAY]: (label: string) =>
    toLocaleDateStringUTC(fromUnixTime(parseInt(label, 10))),
  [ChartTimeGroupings.BY_WEEK]: (_, payload) => {
    const dataDatetime = fromUnixTime(parseInt(payload[0].payload.name, 10));
    const start = toLocaleDateStringUTC(dataDatetime);
    const end = toLocaleDateStringUTC(endOfWeek(dataDatetime));
    return `${start} - ${end}`;
  },
  [ChartTimeGroupings.BY_MONTH]: (_, payload) => {
    const dataDatetime = fromUnixTime(parseInt(payload[0].payload.name, 10));
    return `${formatUTC(dataDatetime, 'MM.YYYY')}`;
  },
};
