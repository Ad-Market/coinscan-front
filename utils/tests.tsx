import { ReactNode } from 'react';
import { faker } from '@faker-js/faker';
import { ApolloClient, ApolloProvider, InMemoryCache } from '@apollo/client';
import { SchemaLink } from '@apollo/client/link/schema';
import { loadSchemaSync } from '@graphql-tools/load';
import { JsonFileLoader } from '@graphql-tools/json-file-loader';
import { addMocksToSchema, IMocks } from '@graphql-tools/mock';
import { IResolvers } from '@graphql-tools/utils';
import { render as rtlRender } from '@testing-library/react';

const schema = loadSchemaSync('./graphql.schema.json', {
  loaders: [new JsonFileLoader()],
});

export function renderWithApolloMocks(
  component: ReactNode,
  mocks: IMocks<IResolvers>
) {
  const mockSchema = addMocksToSchema({
    schema,
    ...mocks,
  });

  const client = new ApolloClient({
    link: new SchemaLink({ schema: mockSchema }),
    cache: new InMemoryCache(),
  });

  return rtlRender(
    <ApolloProvider client={client}>{component}</ApolloProvider>
  );
}
