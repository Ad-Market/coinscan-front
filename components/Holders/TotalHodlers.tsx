import { ApolloError } from '@apollo/client';
// import { TotalHoldersQuery } from '../../generated/graphql';
// import { gql, useQuery } from '@apollo/client';
import { DailyHodlersStatesQuery } from '../../generated/graphql';
import { Loading } from '../Wallets/Wallets.styled';
import Growth from '../Growth';

// const GET_TOTAL_HOLDERS = gql`
//   query TotalHolders {
//     holdersCounters {
//       id
//       count
//     }
//   }
// `;

type TotalHoldersProps = {
  loading?: boolean;
  error?: ApolloError;
  data?: DailyHodlersStatesQuery;
};

function getPercentChange(
  current: number | undefined,
  prev: number | undefined
): number | undefined {
  if (current && prev) {
    const percent = (current / prev) * 100 - 100;
    return percent;
  }

  return;
}

const TotalHodlers = ({ loading, error, data }: TotalHoldersProps) => {
  // const { loading, error, data } =
  //   useQuery<TotalHoldersQuery>(GET_TOTAL_HOLDERS);

  if (loading) return <Loading>Loading...</Loading>;
  if (error) return <div>{error.toString()}</div>;

  const currentHoldersCount = data?.dailyHoldersStates[0].count;
  const oneDayHoldersCount = data?.dailyHoldersStates[1].count;
  const sevenDaysHoldersCount = data?.dailyHoldersStates[6].count;
  const thirtyDaysHoldersCount = data?.dailyHoldersStates[29].count;

  return (
    <Growth
      value={currentHoldersCount}
      oneday={getPercentChange(currentHoldersCount, oneDayHoldersCount)}
      sevendays={getPercentChange(currentHoldersCount, sevenDaysHoldersCount)}
      thirtydays={getPercentChange(currentHoldersCount, thirtyDaysHoldersCount)}
    />
  );
};

export default TotalHodlers;
