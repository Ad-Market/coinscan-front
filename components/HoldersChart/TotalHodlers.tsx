import { useState } from 'react';
import { TotalHoldersQuery } from '../../generated/graphql';
import { gql, useQuery } from '@apollo/client';
import { Loading } from '../Wallets/Wallets.styled';

const GET_TOTAL_HOLDERS = gql`
  query TotalHolders {
    holdersCounters {
      id
      count
    }
  }
`;

const TotalHodlers = () => {
  const { loading, error, data, fetchMore } =
    useQuery<TotalHoldersQuery>(GET_TOTAL_HOLDERS);

  if (loading) return <Loading>Loading...</Loading>;
  if (error) return <div>{error.toString()}</div>;

  return (
    <div>
      <span>Total holders: </span>
      {data?.holdersCounters[0]?.count}
    </div>
  );
};

export default TotalHodlers;
