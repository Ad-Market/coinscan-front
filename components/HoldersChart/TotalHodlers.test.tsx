import { screen } from '@testing-library/react';
import snapshotDiff from 'snapshot-diff';
import { renderWithApolloMocks } from '../../utils/tests';
import TotalHodlers from './TotalHodlers';

test('render total holders with loading pre-screen', async () => {
  const { asFragment } = renderWithApolloMocks(<TotalHodlers />, {
    mocks: {
      HoldersCounter: () => ({ id: 'HOLDERS_COUNTER', count: '997' }),
    },
  });
  const firstRender = asFragment();
  expect(firstRender).toMatchSnapshot('loading');
  await screen.findByText('997');
  expect(snapshotDiff(firstRender, asFragment())).toMatchSnapshot('loaded');
});
