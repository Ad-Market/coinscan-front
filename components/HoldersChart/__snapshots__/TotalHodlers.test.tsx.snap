// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render total holders with loading pre-screen 1`] = `
<DocumentFragment>
  <div
    class="css-13xyho9"
  >
    Loading...
  </div>
</DocumentFragment>
`;

exports[`should render total holders with loading pre-screen 2`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
-   <div
-     class=\\"css-13xyho9\\"
-   >
-     Loading...
+   <div>
+     <span>
+       Total holders: 
+     </span>
+     997
    </div>
  </DocumentFragment>"
`;
